test/cql-pytest/fetch_scylla.py

```diff
@@ -0,0 +1,123 @@
+
+def s3_download(s3, bucket, object_key, out):
+    print(f'Downloading {object_key}... ', end='')
+    if not sys.stdout.isatty():
+        bucket.download_file(object_key, out)
+    else:
+        # When interactive, show download progress:
+        print('')
+        objlen = s3.Object(bucket.name, object_key).content_length
+        downloaded = 0
+        def progress(chunk):
+            nonlocal downloaded
+            downloaded += chunk
+            print('\33[2K\r%02d%%  %d MB' % (int(downloaded*100/objlen), int(downloaded/1024/1024)), end='')
+        bucket.download_file(object_key, out, Callback=progress)
+        print('')
+    print('done.')
+
+# Download Scylla release "release" (e.g., 5.4, 5.4.7, 2023.1, 2023.1.7)
+# into a subdirectory of directory "dir" named by the specific release
+# downloaded (e.g., 5.4.7) and return the ready-to-use scylla executable
+# (actually, a shell-script wrapper) in that directory - or raise exception
+# if this release cannot be found.
+# If the same download was already done in the past, discover this as quickly
+# as possible (a 3-component release like 5.4.7 doesn't even require a network
+# lookup, a 2-component release like 5.4 does) and don't download again.
+def download_scylla(release, dir):
+    scylla_arch = 'x86_64'
+
+    v = parse_version(release)
+    # If release has 3 components (e.g., 5.4.7) we know exactly which version
+    # we are looking for, and can check if we already downloaded it without
+    # even fetching the list of versions from s3:
+    if len(v) == 3 and os.path.exists(dir + '/' + release + '/scylla_wrapper'):
+        print(f'Already downloaded, in {dir}/{release}')
+        return dir + '/' + release + '/scylla_wrapper'
+    major = f'{v[0]}.{v[1]}'
+    bucket = f'downloads.scylladb.com'
+    scylla_arch_string = '.' + scylla_arch + '.'
+    if int(v[0]) >= 2023: # Enterprise release (new organization)
+        prefix = f'downloads/scylla-enterprise/relocatable/scylladb-{major}/scylla-enterprise-{release}'
+    elif int(v[0]) > 2000: # Enterprise release (old organization)
+        prefix = f'downloads/scylla-enterprise/relocatable/scylladb-{major}/scylla-enterprise-{scylla_arch}-package-{release}'
+        scylla_arch_string = '' # arch already restricted by prefix
+    elif [int(v[0]),int(v[1])] <= [4,5]: # Open source (very old organization)
+        prefix = f'downloads/scylla/relocatable/scylladb-{major}/scylla-package-{release}'
+        scylla_arch_string = '' # no arch (only x86)
+    elif [int(v[0]),int(v[1])] <= [5,1]: # Open source (old organization)
+        prefix = f'downloads/scylla/relocatable/scylladb-{major}/scylla-{scylla_arch}-package-{release}'
+        scylla_arch_string = '' # arch already restricted by prefix
+    else: # Open source (new organization)
+        prefix = f'downloads/scylla/relocatable/scylladb-{major}/scylla-{release}'
+    # This prefix has many different packages belonging to all releases in
```

Comment:
This looks super complicated. Why not use the container image? Or require the user to run whatever they want on the side?

Reply:
It is complicated because the person who maintains downloads.scylladb.com made this complicated by changing the naming of the files over the years.

But the beauty of writing a script is that all the complexity falls on the person who wrote it (me) - the user just gets version 2024.1.17 (or whatever) as efficiently as possible.

You are suggesting the **wrong** (in my opinion) tradeoff - instead of me working hard(ish) to write this script, let me just work less and use docker (which already works) - but the user's experience will suffer: The docker images are significantly bigger making them slower to download and significantly more taxing on your disk to cache locally. Running a docker image with all the garbage it runs is also significantly slower than starting Scylla directly the way that test/cql-pytest/run currently starts Scylla.

Additionally, test/cql-pytest/run already knows how to run a Scylla executable in a certain location, not a docker image. Making it run a docker image is very much possible (I also considered that option) but also requires work.

Regarding the last option you mentioned - "require the user to run whatever they want on the side" - this is indeed one of the strong suites of the cql-pytest framework - it allows you to run Scylla (whichever Scylla you want, however you want) and then run tests against that. This is exactly what I've been doing in the past. This patch is about making this easier - let the test/cql-pytest/run convenience script run Scylla for you.

Reply:
Well, I don't mind merging it, but my expectation is that it will randomly break from time to time.

Reply:
It may indeed break. The most fragile part isn't the download part that you pointed to (I assume that we won't go back to the download directory and rename files) but rather the part where we need to figure out what are the correct configuration options to pass to older versions. This part is the ugliest. But I do think it's worth it - fixing (I will do it) this script once in a few months that it breaks, and everyone can benefit from the working script without needing to understand the details.